frontend/src/pages/MatchingGame.jsx

```diff
@@ -0,0 +1,118 @@
+import React, { useState, useEffect } from "react";
+
+const MatchingGame = ({ test, setMode, setSelectedTest }) => {
+    const [shuffledQuestions, setShuffledQuestions] = useState([]);
+    const [shuffledAnswers, setShuffledAnswers] = useState([]);
+    const [selectedQuestion, setSelectedQuestion] = useState(null);
+    const [selectedAnswer, setSelectedAnswer] = useState(null);
+    const [matchedPairs, setMatchedPairs] = useState([]);
+    const [errorMessage, setErrorMessage] = useState("");
+
+    useEffect(() => {
+        if (test.questions.length < 3) {
+            alert("Matching Game requires at least 4 questions.");
+            setMode(null);
+            setSelectedTest(null);
+        } else {
+            setShuffledQuestions([...test.questions].sort(() => Math.random() - 0.5));
+            setShuffledAnswers([...test.questions].sort(() => Math.random() - 0.5));
+        }
+    }, [test, setMode, setSelectedTest]);
+
+    const handleQuestionClick = (question) => {
+        setSelectedQuestion(question);
+        setErrorMessage("");
+    };
+
+    const handleAnswerClick = (answer) => {
+        setSelectedAnswer(answer);
+        setErrorMessage("");
+    };
+
+    const checkMatch = () => {
+        if (selectedQuestion && selectedAnswer) {
+            if (selectedQuestion.answer === selectedAnswer.answer) {
+                setMatchedPairs([...matchedPairs, selectedQuestion.question]);
+                setSelectedQuestion(null);
+                setSelectedAnswer(null);
+            } else {
+                setErrorMessage("Incorrect match! Try again.");
+            }
+        }
+    };
+
+    const isMatched = (item) =>
+        matchedPairs.includes(item.question || item.answer);
+
+    return (
+        <div className="bg-[#0f172a] min-h-screen flex flex-col items-center py-12 px-6 text-white">
+            <h1 className="text-4xl font-bold mb-8">Matching Game</h1>
+
+            <div className="grid grid-cols-4 gap-6 w-full max-w-4xl">
+                {/* Pytania */}
```

Comment:
Let's change the comment to English.
```suggestion
                {/* Questions */}
```